external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala

```diff
@@ -251,32 +274,53 @@ private[kafka010] case class InternalKafkaConsumer(
       untilOffset: Long,
       pollTimeoutMs: Long,
       failOnDataLoss: Boolean): ConsumerRecord[Array[Byte], Array[Byte]] = {
-    if (offset != nextOffsetInFetchedData || !fetchedData.hasNext()) {
-      // This is the first fetch, or the last pre-fetched data has been drained.
+    if (offset != nextOffsetInFetchedData) {
+      // This is the first fetch, or the fetched data has been reset.
       // Seek to the offset because we may call seekToBeginning or seekToEnd before this.
       seek(offset)
       poll(pollTimeoutMs)
+    } else if (!fetchedData.hasNext) {
+      // The last pre-fetched data has been drained.
+      if (offset < offsetAfterPoll) {
+        // Offsets in [offset, offsetAfterPoll) are missing. We should skip them.
+        resetFetchedData()
+        throw new MissingOffsetException(offset, offsetAfterPoll)
```

Comment:
So MissingOffsetRange is only used to signal that some offset may be missing due to control messages and nothing else. And the higher function (i.e. `get`) just handles it by resetting the fetched offsets. Why not let this `fetchData` method handle the situation instead of creating a new exception just for this?